File: Tarea11/src/Metodo2(ast)/Operacion.ts
export class Operacion{

    public valor : string;
    public izquierda : Operacion | null;
    public derecha : Operacion | null;
    public temp : number = 1;

    constructor(valor : string , izq : Operacion, der : Operacion ){
        this.valor = valor;
        this.izquierda = izq;
        this.derecha = der;
     }

    public traducir() : string{
        if(this.izquierda != null && this.derecha != null){
            let izq = this.izquierda?.traducir();
            //Pasar al sub arbol derecho la profundidad del sub arbol izquierdo
            this.derecha.temp = this.izquierda.temp;
            let der = this.derecha?.traducir();

            let operador = this.valor;
            this.valor = this.newTemp();
            return `${izq}\n${der}\n${this.valor} = ${this.izquierda.valor} ${operador} ${this.derecha.valor}`;
        }
        return "";
    }

    public newTemp():string{
        //Verificar si la profundidad actual heredo una profundidad mas baja que la de los hijos
        if(this.temp < this.izquierda?.temp! || this.temp < this.derecha?.temp!){
            if(this.izquierda?.temp! >= this.derecha?.temp!){
                this.temp = this.izquierda?.temp!;
            }else{
                this.temp = this.derecha?.temp!;
            }
        }
        let str_temp = `t${this.temp}`;
        this.temp++;
        return str_temp;
    }

};

File: Tarea12/src/operacion.py
class Operacion() :
    def __init__(self, izq, der, operador) :
        self.izq = izq
        self.der = der
        self.operador = operador
        self.temporal = 1
    
    def new_temp(self) :
        if self.temporal < self.izq.temporal or self.temporal < self.der.temporal:
            if self.izq.temporal >= self.der.temporal :
                self.temporal = self.izq.temporal
            else :
                self.temporal = self.der.temporal
        str_temp = f"t{self.temporal}"
        self.temporal+=1
        return str_temp

    def traducir(self):
        print(self)
        if self.izq is not None and self.der is not None:
            v_izq = self.izq.traducir()
            self.der.temporal = self.izq.temporal
            v_der = self.der.traducir()
            op = self.operador
            self.operador = self.new_temp()
            return f"{v_izq}\n{v_der}\n{self.operador}={self.izq.operador} {op} {self.der.operador}"
        else:
            return ""
        
